tensorflow/python/keras/integration_test/gradient_checkpoint_test.py

```diff
@@ -0,0 +1,160 @@
+
+
+def _limit_gpu_memory():
+  """Helper function to limit GPU memory for testing  """
+  gpus = tf.config.experimental.list_physical_devices('GPU')
+  if gpus:
+    try:
+      tf.config.experimental.set_virtual_device_configuration(
+          gpus[0], [
+              tf.config.experimental.VirtualDeviceConfiguration(
+                  memory_limit=1024)
+          ])
+    except RuntimeError as e:
+      print(e)
+
+
+def _get_dummy_data(img_dim, n_channels, batch_size):
+  inputs = tf.ones([batch_size, img_dim, img_dim, n_channels])
+  labels = tf.ones([batch_size], dtype=tf.int64)
+  return inputs, labels
+
+
+def _train_no_recompute(n_steps):
+  """Trains a single large model without gradient checkpointing."""
+  _limit_gpu_memory()
+  img_dim, n_channels, batch_size = 256, 1, 4
+  x, y = _get_dummy_data(img_dim, n_channels, batch_size)
+  model = _get_big_cnn_model(img_dim,
+                             n_channels,
+                             num_partitions=3,
+                             blocks_per_partition=2)
+  optimizer = optimizers.SGD()
+  losses = []
+  tr_vars = model.trainable_variables
+  for _ in range(n_steps):
+    with tf.GradientTape() as tape:
+      logits = model(x)
+      loss = _compute_loss(logits, y)
+      losses.append(loss)
+    grads = tape.gradient(loss, tr_vars)  # tr_vars
+    optimizer.apply_gradients(zip(grads, tr_vars))
+    del grads
+  return losses
+
+
+def _train_with_recompute(n_steps):
+  """Trains a single large model with gradient checkpointing using tf.recompute_grad."""
+  _limit_gpu_memory()
+  img_dim, n_channels, batch_size = 256, 1, 4
+  x, y = _get_dummy_data(img_dim, n_channels, batch_size)
+  # This model is the same model as _get_big_cnn_model but split into 3 parts.
+  models = _get_split_cnn_model(img_dim,
+                                n_channels,
+                                num_partitions=3,
+                                blocks_per_partition=2)
+  model1, model2, model3 = models
+  # Apply gradient checkpointing to the submodels using tf.recompute_grad.
+  model1_re = tf.recompute_grad(model1)
+  model2_re = tf.recompute_grad(model2)
+  model3_re = tf.recompute_grad(model3)
+  optimizer = optimizers.SGD()
+  tr_vars = model1.trainable_variables + model2.trainable_variables + model3.trainable_variables
+  losses = []
+  for _ in range(n_steps):
+    with tf.GradientTape() as tape:
+      logits1 = model1_re(x)
+      logits2 = model2_re(logits1)
+      logits3 = model3_re(logits2)
+      loss = _compute_loss(logits3, y)
+      losses.append(loss)
+      grads = tape.gradient(loss, tr_vars)  # tr_vars
+      optimizer.apply_gradients(zip(grads, tr_vars))
+      del grads
+  return losses
+
+
+class GradientCheckpointTest(tf.test.TestCase):
+
+  def test_raises_oom_exception(self):
+    with self.assertRaises(Exception) as context:
+      _train_no_recompute(1)
+    self.assertTrue(
+        context.exception.__class__.__name__ == 'ResourceExhaustedError')
```

Comment:
I am scared of tests that check for actual OOMs as they are sensitive to things like what specific device it's run on, etc.

I see two options here:
 1. use a virtual GPU device with fixed limited memory
 2. (my preferred alternative) use side effects to detect whether things are recomputed. For example if the forward pass does variable.assign_add(1) we can count how many times it ran

Reply:
I am using a virtual GPU if you look at line 76 inside the _limit_gpu_memory() function. Maybe I should bazel tag these tests to run only on a virtual GPU? I actually like the OOM test since it demonstrates and confirms that the grad checkpointing is functional. Otherwise it might remain silent and broken like it has been in TF 2.x all along. But happy to roll it back and add your preferred option 2 if you feel this is risky.

Reply:
Ah I missed it. This is good then. Thanks!